### DISCUSSION TITLE
Link to custom sqlite library

### DISCUSSION THREAD

**Discussion by headmastersquall** [rusqlite/rusqlite#1172] [Questions] [answered]
I have a project that needs to link to SQLite with SEE. I have the compiled binary in sqlite/libsqlite3.so along with sqlite/sqlite3.h and sqlite/sqlite3.c. I am trying to build my project with:

`SQLITE3_LIB_DIR=./sqlite SQLITE3_INCLUDE_DIR=./sqlite cargo build`

But it still links to the system library:

```
ldd target/debug/server
	linux-vdso.so.1 (0x00007fff62549000)
	libsqlite3.so.0 => /lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007f7a2cb59000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7a2cb3e000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7a2cb1b000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7a2c9cc000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f7a2c9c6000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7a2c7d4000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f7a2d858000)
```

How can I get my project to link against my provided library? Thank you.

**Comment by gwenn** [comment]
#731
> You can take a look at target/debug/build/libsqlite3-sys-<hash>/output, which contains what our build.rs told cargo to link to (note that each rebuild will produce a folder with this output file and without -- the one without is compilation info for the build.rs), as well as inspecting what cargo build --verbose.

> I think I see what's happening. I need to use LD_LIBRARY_PATH to point to the /usr/lib/sqlite-replication forked sqlite3 lib at runtime or else dqlite will load the non-forked sqlite3 from my system.

...

**Comment by headmastersquall** [comment, marked as answer]
Finally figured it out. This was tricky, but simple at the same time. Here is what I had to do to get my project to build with SQLite that includes SEE.

```
# First, the SQLite source must be compiled to a static library
OUT_DIR=system-lib
rm -rf $OUT_DIR
mkdir -p $OUT_DIR

# Extracted sources files from SQLite-SEE
cd see-sources

echo "Buildig libsqlite3.so"
gcc -fPIC -shared -o ../$OUT_DIR/libsqlite3.so sqlite3-see.c

echo "Building static library"
gcc -c sqlite3-see.c -o libsqlite3.o
ar -crs libsqlite3.a libsqlite3.o
rm libsqlite3.o
mv libsqlite3.a ../$OUT_DIR
cp sqlite3.h ../$OUT_DIR
```

```
# system-lib will contain these files:
libsqlite3.so
libsqlite3.a
sqlite3.h
```

```
# Then, the rust project can be built by setting these environment variables
export SQLITE3_LIB_DIR=system-lib
export SQLITE3_INCLUDE_DIR=system-lib
export SQLITE3_STATIC=1
cargo build --release
```
